---
title: "Merging taxonomy info, AMR genes detected with BLASTn and KMA against Resfinder database and mobile genetic elements detected in mobileOG database"
output: html_notebook
---
```{r setup}

library(dplyr)
library(stringr) #for str_extract function
library(jsonlite) #for reading in json file
library(tidyr) #for use with replace_na

packageVersion("dplyr")
```

#load in taxa results (stringent)

load in all the taxonomy results from each run, merging them into one file
```{r}
load("cvn1_stringent_2025_clean4") #only barcodes 2,5  and 8 from this run
load("cvn2_stringent_2025_clean4")#only barcodes barcode3","barcode4","barcode5","barcode6","barcode7","barcode8","barcode9","barcode22" kept
load("cvn3_stringent_2025_clean4") #only barcodes "barcode1","barcode2","barcode10" kept
load("stringent_2025_gm1")
load("stringent_2025_gm2")
load("stringent_2025_gm4") #only barcodes 5-20 kept from gm3 run

merged_all_runs <- bind_rows(cvn1_stringent_2025_clean4,cvn2_stringent_2025_clean4,cvn3_stringent_2025_clean4,stringent_2025_gm1,stringent_2025_gm2,stringent_2025_gm4)

save(merged_all_runs, file = "merged_all_runs") #saving the file for quick reloading later on

load("merged_all_runs")

all_runs_no_controls <- merged_all_runs %>%
                        filter(!sample_name %in% c("HMW_POS_CTRL", "LOG_DNA_POS_CTRL","LOG_POS_CTRL", "16S_POS_CTRL", "NEG_CTRL", "NEG_EXT_CTRL"))

save(all_runs_no_controls, file = "all_runs_no_controls") #saving the file for quick reloading later on

load("all_runs_no_controls")
```


Load in merged resfinder results:

```{r}
gm_all_blastn_resfinder <-  read.table("gut_microbiome_all_blastn_resfinder.txt", sep="",  header = FALSE) 

colnames(gm_all_blastn_resfinder) <- c("subject.sci.names","query.acc","subject.acc", "percent.identity", "coverage", "alignment.length", "mismatches","gap.opens", "gaps", "subject.length", "query.length", "q.start", "q.end", "s.start", "s.end", "evalue", "subject.tax.ids")


#*Add in a alignment length to query/subject length proportion column to further illustrate how much of a match we are seeing**

gm_all_blastn_resfinder_filtered <- gm_all_blastn_resfinder%>%
     mutate(subject.alignment.frac = alignment.length/subject.length, 
         query.alignment.frac = alignment.length/query.length) %>%
           filter(percent.identity >= 80, subject.alignment.frac >= .6) #814 hits
```

```{r}

#merge the resfinder data with taxa info

gm_resfinder_blastn_2025_taxa <- left_join(all_runs_no_controls, gm_all_blastn_resfinder_filtered , by = "query.acc", relationship = "many-to-many")

save(gm_resfinder_blastn_2025_taxa, file = "gm_resfinder_blastn_CARD_2025_taxa") #saving the file for quick reloading later on

load("gm_resfinder_blastn_CARD_2025_taxa")
```

load in mge

```{r}

#load in mge


   
gm_mge <- read.csv(file = "gut_microbiome_all_mge.csv",  header = TRUE)

#change the name of Specific.Contig to query.acc so that I can merge the amr_filters_CARD and the grey_seal_mge datasets together using this linker column, which is the query name, but just with different headings on both datasets:

gm_mge$query.acc <- gm_mge$Specific.Contig


#merge the mge results

gm_blastn_resfinder_2025_taxa_mge <- left_join(gm_resfinder_blastn_2025_taxa, gm_mge, by = "query.acc", relationship = "many-to-many")

save(gm_blastn_resfinder_2025_taxa_mge, file = "gm_blastn_resfinder_2025_taxa_mge") #saving the file for quick reloading later on
load("gm_blastn_resfinder_2025_taxa_mge") ##this is the dataset that has EVERYTHING - The blastn amr gene hit, if its associated awith a bacterial spp from the sample and whether this amr gene is a mge

nrow(gm_blastn_resfinder_2025_taxa_mge) #472708 resfinder amr hits, but this also includes taxa with no hits, mge with no hits etc

```

```{r}
#filter this dataset, gm_all_blastn_resfinder_filtered  further, because there are multiple hits of the same read and we only want one hit per read:

#remember you are filtering by the .y not .x evalue, alignment length etc because you are filtering the reads based on the amr results (.y) not the species results (.x)


gm_blastn_resfinder_2025_taxa_mge_cleaned <- gm_blastn_resfinder_2025_taxa_mge %>%
group_by (query.acc, subject.acc.y) %>% # FILTERING STEP 1: by grouping the read with saccession, if the same read hits to multiple samr genes, all these genes are retained for that read
            
               mutate(evalue_nonzero = ifelse(evalue.y == 0, 1, evalue.y), 
                      relative_evalue = ifelse(evalue.y == 0, 1, min(evalue_nonzero)/evalue_nonzero)) %>%
              mutate(relative_alignment_length = alignment.length.y/max(alignment.length.y)) %>%
              mutate(alignment_count = n()) %>% #taxon alignment count within a read 
              
              ungroup() %>%
              group_by(query.acc) %>%
             
              mutate(relative_count = alignment_count/n()) %>% #proportion of read alignments to a taxon
               mutate(weight_score = relative_evalue*relative_alignment_length*(percent.identity.y/100)) %>%
             filter(weight_score == max(weight_score)) %>%
              filter(relative_count == max(relative_count)) %>%
              slice_sample(n=1) %>%
              ungroup() #28 hits for resfinder genes %>%
              

save(gm_blastn_resfinder_2025_taxa_mge_cleaned, file = "gm_blastn_resfinder_2025_taxa_mge_cleaned") #saving the file for quick reloading later on
load("gm_blastn_resfinder_2025_taxa_mge_cleaned") ##give to ELEANOR, blastN RESULTS S AUREUS THAT HAVE MGE

##resfinder amr genes on mge:

gm_blastn_resfinder_2025_taxa_mge_hits_cleaned <- gm_blastn_resfinder_2025_taxa_mge_cleaned %>%
                        filter (!Sequence.Title == "NA")# this dataset represents amr genes on identified species, and all the genes are on mobile genetic elements,. 3hits
write.csv(gm_blastn_resfinder_2025_taxa_mge_hits_cleaned,"For_Eleanor_gm_blastn_resfinder_2025)taxa_mge.csv")



#lets load in the info from resfinder about amr gene names and resistance info:

resfinder_amr_info <-  read.delim("phenotypes.txt", header = TRUE) 

resfinder_amr_info <- resfinder_amr_info %>%
                      rename(subject.acc.y = Gene_accession.no.)
  
  
gm_blastn_resfinder_2025_taxa_mge_cleaned  <- left_join(gm_blastn_resfinder_2025_taxa_mge_cleaned, resfinder_amr_info, by = "subject.acc.y", relationship = "many-to-many") 
                    
#for thesis

##lets get a breakdown of the amr gene names on bacterial spp detected with blastn on CARD for protein homolog hits only:
gm_gene_names_resfinder_blastn_clean_ph <- gm_blastn_resfinder_2025_taxa_mge_cleaned %>%
  select(subject.acc.y, Class, Mechanism.of.resistance, species.x,percent.identity.x, evalue.x,Major.mobileOG.Category) %>%
  group_by(subject.acc.y, Class, Mechanism.of.resistance, species.x, Major.mobileOG.Category) %>%
  summarise(count = n()) %>%
  mutate(prop =prop.table(count)) %>%
  ungroup() %>%
  arrange(desc(prop)) %>%
 mutate(Accession.no. = str_extract(subject.acc.y, "[^_]+$")) #pulling out the accession names (which are the strings after the last underscore in the subject.query.acc column and outputting it to a new column Acesssion.no, which matches the column name for accession numbers outputted by kma results below, so i can merge ion this column)


save(gm_gene_names_resfinder_blastn_clean_ph , file = "gm_gene_names_resfinder_blastn_clean_ph ") #saving the file for quick reloading later on
load("gm_gene_names_resfinder_blastn_clean_ph")

 
write.csv(gm_gene_names_resfinder_blastn_clean_ph,"gm_gene_names_blastn_resfinder.csv")

#the blastn resfinder reuslts are filtered ot retain only reads that have 60% coverage or more across and amr gene and the percent identity to that amr gene is 80% or more


```


KMA Results:

```{r}

#load in kma file

gm_resfinder_kma <- read.csv("gm_all_kma_summary.csv") #note that i manually made this table in excel by searching in the concatenated kma file for all gut microbiome chapter samples and compilling them into this excel file



#rename contig as query.acc as these are the read names

gm_resfinder_kma$query.acc <- gm_resfinder_kma$Contig



#merge the resfinder data with taxa info

gm_resfinder_kma_taxa <- left_join(gm_resfinder_kma , all_runs_no_controls , by = "query.acc", relationship = "many-to-many")

save(gm_resfinder_kma_taxa, file = "gm_resfinder_kma_taxa") #saving the file for quick reloading later on
load("gm_resfinder_kma_taxa")


#merge the resfinder kma/taxa dat with mge data

gm_resfinder_kma_taxa_mge <- left_join(gm_resfinder_kma_taxa, gm_mge, by = "query.acc", relationship = "many-to-many")

save(gm_resfinder_kma_taxa_mge, file = "gm_resfinder_kma_taxa_mge") #saving the file for quick reloading later on
load("gm_resfinder_kma_taxa_mge") ##this is the dataset that has EVERYTHING - The blastn amr gene hit, if its associated awith a bacterial spp from the sample and whether this amr gene is a mge

#remember you are filtering by the .y not .x evalue, alignment length etc because you are filtering the reads based on the amr results (.y) not the species results (.x)



#note the filtering is a little different to above as we didnt run a blast analysis, so dont have evalues, alignment lengths etc for the amr hits. we just have what the kma algorithm outputted - which does contain %id and coverage, so we can filter to reatain only one read that has both the highest poertcent id and coverage across the amr gene found: THIS FILTERING IS CURRENTLY NOT WORKING AND BOTH READS

gm_resfinder_kma_taxa_mge$Identity <- as.numeric(gm_resfinder_kma_taxa_mge$Identity)
gm_resfinder_kma_taxa_mge$Coverage <- as.numeric(gm_resfinder_kma_taxa_mge$Coverage)

gm_resfinder_kma_taxa_mge_cleaned <- gm_resfinder_kma_taxa_mge %>%
group_by (query.acc, Identity, Coverage, Resistance.gene) %>% #im adding resistance gene here as when i didnt do this, i was still getting a single read hitting to two different amr genes being retained, but again i only want a single hit per read. the amr genes must be very genetically similar (it was aad and tetL i was looking at, because both hits for the same read were over 99% id and 98% coverage for both genes. But lets be really strict and only keep the best hit)
            
               mutate(weight_score2 = Coverage * Identity) %>%
             filter(weight_score2 == max(weight_score2)) %>%
            #  filter(relative_count == max(relative_count)) %>%
              slice_sample(n=1) %>%
              ungroup() %>% #120 hits for resfinder genes
             rename(subject.acc.y = Resistance.gene) %>%
              rename(Class = Resistance)
              #have to rename them to match blastn resfinder results to merge later
  
  
save(gm_resfinder_kma_taxa_mge_cleaned, file = "gm_resfinder_kma_taxa_mge_cleaned") #saving the file for quick reloading later on
load("gm_resfinder_kma_taxa_mge_cleaned") 

##JUST PULL OUT ONES ON MGE:

gm_resfinder_kma_mge <- gm_resfinder_kma_taxa_mge_cleaned %>%
                          filter (!Sequence.Title == "NA")



write.csv(gm_resfinder_kma_mge,"For_Eleanor_gm_resfinder_kma_mge.csv")



#breakdown of kma results
gm_gene_names_resfinder_kma <- gm_resfinder_kma_taxa_mge_cleaned %>%
  select(subject.acc.y, Class, Accession.no., species.x, Major.mobileOG.Category) %>%
  group_by(subject.acc.y, Class, Accession.no., species.x,Major.mobileOG.Category) %>%
  summarise(count = n()) %>%
  mutate(prop =prop.table(count)) %>%
  ungroup() %>%
  arrange(desc(prop))
 

                               
                                
write.csv(gm_gene_names_resfinder_kma,"gm_gene_names_resfinder_kma.csv")



##merge this with blastn resfinder counts:

shared_genes_all_resfinder<- gm_gene_names_resfinder_kma  %>%
                    inner_join(gm_gene_names_resfinder_blastn_clean_ph, by = c("Accession.no.", "Class", "species.x", "Major.mobileOG.Category"))

#pull out columns we care about into a final count table:

shared_genes_all_resfinder_final <-shared_genes_all_resfinder %>%
                                   
                                    select(subject.acc.y.x, Class, Mechanism.of.resistance, species.x, count.x, count.y, Major.mobileOG.Category)

write.csv(shared_genes_all_resfinder_final,"gm_gene_names_resfinder_final.csv")
                                

```


